torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -223,11 +225,41 @@ struct DeviceTypePropagationPass {
     return changed;
   }
 
-  /*
-  bool processLoop(Node* n) {
+  bool applyTensorProps(
+      const at::ArrayRef<Value*>& src,
+      const at::ArrayRef<Value*>& dst) {
+    TORCH_INTERNAL_ASSERT(src.size() == dst.size());
+    bool changed = false;
+    for (int i = 0; i < dst.size(); i++) {
+      auto src_type = src[i]->type()->cast<TensorType>();
+      changed |= setDeviceType(dst[i], src_type->device());
+    }
+    return changed;
+  }
+
+  void processLoop(Node* node) {
     GRAPH_DEBUG("processLoop");
+    LoopView l(node);
+
+    applyTensorProps(l.carriedInputs(), l.bodyCarriedInputs());
+
+    int iter = 0;
+    for (; iter < 4; iter++) {
+      processBlock(l.bodyBlock());
+      bool inputs_changed = mergeAndApplyTensorProps(
+          l.carriedInputs(), l.bodyCarriedOutputs(), l.bodyCarriedInputs());
+      if (!inputs_changed) {
+        break;
+      }
+    }
+    TORCH_INTERNAL_ASSERT(
```

Comment:
We shouldn't be throwing errors in analysis passes - they should just not succeed gracefully

Reply:
Do you want me to mark all outputs as `c10::nullopt` device, or do you have a different ideas as to how to "not succeed gracefully" ?

Reply:
Yes, just set to `c10::nullopt`